### PR TITLE
Gem2s loading UI

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-751

#### Link to staging deployment URL 

https://ui-agi-ui257.scp-staging.biomage.net/experiments/7d9318d157a6b17d3d6caec23a25cc86/data-processing

#### Links to any Pull Requests related to this

N.A.

#### Anything else the reviewers should know about the changes here

- Does not appear on staging yet because relevant states are not hooked up.
- Staging test to ensure the changes does not interfere with current state.

# Changes
### Code changes

- Created additional 2 screens to display info for GEM2S not run and GEM2error.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [X] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### Photos

No data
image

Running
image

Error
image

### CODE REVIEW

**Comment by aerlaut** [comment]
@vickymorrison can you please take a look the UI? The changes can not yet be seen via the staging link, as gem2s is not yet implemented. I've taken captures of the screen for the 3 states of GEM2S: no data uploaded, GEM2S is running, and there is GEM2S error. Do you have any comments on them?

File: src/components/ContentWrapper.jsx [hunk 1]
```diff
@@ -39,22 +40,27 @@ const ContentWrapper = (props) => {
   const route = router?.route || '';
 
   const {
-    loading: pipelineLoading,
-    error: pipelineError,
-    status: workerAndPipelineStatus,
+    loading: backendLoading,
+    error: backendError,
+    status: backendStatus,
```
Comment (ivababukova) [inline src/components/ContentWrapper.jsx:45]: I am not seeing in this PR where actually `pipelineError`, `pipelineLoading`, ... etc got renamed to `backend...` -- is it part of a different PR?

Comment (ivababukova) [reply src/components/ContentWrapper.jsx:45]: nevermind, it's fine, I just got confused

File: src/components/ContentWrapper.jsx [hunk 2]
```diff
@@ -238,7 +257,7 @@ const ContentWrapper = (props) => {
   }) => {
     const noExperimentDisable = !experimentId ? disableIfNoExperiment : false;
     const pipelineStatusDisable = disabledByPipelineStatus && (
-      pipelineError || pipelineRunning || pipelineRunningError
+      backendError || pipelineRunning || pipelineRunningError
```
Comment (ivababukova) [inline src/components/ContentWrapper.jsx:260]: is `backendError` going to be true if either `pipeline` or `gem2s` had an error?

File: src/components/GEM2SLoadingScreen.jsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
+import React from 'react';
+import {
+  Result, Button, Progress, Row, Col, Typography, Space,
+} from 'antd';
+import Link from 'next/link';
+import PropTypes from 'prop-types';
+
+const { Title, Text } = Typography;
+
+const GEM2SLoadingScreen = ({ pipelineStatus, completedSteps, steps }) => {
```
Comment (ivababukova) [inline src/components/GEM2SLoadingScreen.jsx:10]: This component is mostly the same as `PipelineRedirectToDataProcessing`, apart from the specific message strings and other small details. Have you looked at reusing that component? Or make a new component called `Bac kendLoadingScreen` and use it for both gem2s and data processing?

**Comment by vickymorrison** [comment]
-For 'No Data', when would the user see this screen? When the user logs in, does not upload data and then navigates to either Data Processing/Exploration/Plots modules?
-On the 'Running' screen, 'computing metrics' doesn't make much sense to the user. I don't think it is needed. Can this text be removed? Having the % progress bar and the 'We're launching your analysis...' is enough.
-On the 'Error' screen, if the primary action is for the user to try again then it would be great to have a 'Retry' button directly on this screen (rather than having to navigate back to Data Management and then click on Launch Analysis again). The alternative option to 'Retry' could be to 'Return to Data Management'.

File: src/components/GEM2SLoadingScreen.jsx (hunk 3 above)
Comment (aerlaut) [reply src/components/GEM2SLoadingScreen.jsx:10]: Yes, actually it's very similar. I'll try to refactor that component to make it cover both use cases

File: src/components/ContentWrapper.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/ContentWrapper.jsx:260]: This is actually for showing if there are errors when getting (loading) the backend status. Renamed to `errorLoadingBackend` to make it clearer.

**Comment by aerlaut** [comment]
-For 'No Data', when would the user see this screen? When the user logs in, does not upload data and then navigates to either Data Processing/Exploration/Plots modules?
-On the 'Running' screen, 'computing metrics' doesn't make much sense to the user. I don't think it is needed. Can this text be removed? Having the % progress bar and the 'We're launching your analysis...' is enough.
-On the 'Error' screen, if the primary action is for the user to try again then it would be great to have a 'Retry' button directly on this screen (rather than having to navigate back to Data Management and then click on Launch Analysis again). The alternative option to 'Retry' could be to 'Return to Data Management'.

'No Data' -> This screen will appear if GEM2S has not been run. This will happen if the user directly accesses `/data-processing` without ever launching GEM2S step (i.e. direct access via URL).

'Running' -> I was thinking of having a series of information that would inform the user what step of GEM2S the pipeline is currently running. 'Computing metrics' was one of them. These are the steps :

``` 
const gem2sStepsInfo = [
    'Downloading sample files',
    'Preprocessing samples',
    'Computing metrics',
    'Converting samples',
    'Preparing experiment',
    'Uploading completed data',
  ];
```

So if GEM2S is currently on the first step, it would show 'Downloading sample files', and so on. What do you think? Is this helpful?

'Error' -> Currently we have no way to inform the user what error happened. I don't think retrying GEM2S directly without any changes to the input files would help. Ideally, the user would be informed of the error and so they can make changes to the samples (i.e. reupload/change files) before trying to launch GEM2S again.

**Comment by vickymorrison** [comment]
No data -> Makes sense. It should cover the user landing on data processing, data exploration and plots & tables via direct URL.
Is "Let's upload your data" the best text? Technically they have already uploaded their data but not yet launched the analysis. Perhaps "Let's upload and pre-process your data." would be better.

Running -> OK I understand. I am not convinced it is very helpful to the user but I suppose in the future this might help us understand more about any errors that occur during this process. 

Error -> that makes sense.

**Review by ivababukova — APPROVED** [review]
codewise: lgtm

**Comment by aerlaut** [comment]
> Is "Let's upload your data" the best text? Technically they have already uploaded their data but not yet launched the analysis. Perhaps "Let's upload and pre-process your data." would be better.

Changed accordingly to "Let's upload and pre-process data your data."

**Review by vickymorrison — APPROVED** [review]

**Pull request merged**